### PR TITLE
feat: implement custom headers

### PR DESCRIPTION
**Pull request by sequensucks**
## Description

Add custom headers for all queries and mutations
## Related Issue
re #1364

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: packages/theme/composables/useConfig/index.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ export function useConfig(): UseConfigInterface {
     Logger.debug('useConfig/load');
 
     try {
-      const { data } = await app.$vsf.$magento.api.storeConfig(params?.customQuery ?? null);
+      const { data } = await app.$vsf.$magento.api.storeConfig(params?.customQuery ?? null, params?.customHeaders ?? null);
```
Comment (bartoszherba) [inline packages/theme/composables/useConfig/index.ts:27]: Could you do this everywhere?

```const def = {test: true};
const fun = (val = {}) => {
 console.log({...def, ...val});
}
fun(def?.doesNotExistsProp);
```
Yields 
```
{
  test: true
}
```
So the ?? null is redundant in all cases.

```suggestion
      const { data } = await app.$vsf.$magento.api.storeConfig(params?.customQuery, params?.customHeaders);
```

**Review by bartoszherba — CHANGES_REQUESTED** [review]

**Comment by bartoszherba** [comment]
@sequensucks Please, add guide for the custom headers in the documentation `docs/guide/custom-headers.md`

File: packages/api-client/src/api/getHeaders.ts [hunk 2]
```diff
@@ -1,9 +1,10 @@
...
 
-export default function getHeaders(context: Context) {
+export default function getHeaders(context: Context, customHeaders: Record<string, string> = {}) {
   const { getCustomerToken, getStore, getCurrency } = context.config.state;
 
   return {
+    ...customHeaders,
     ...(getCustomerToken() && { Authorization: `Bearer ${getCustomerToken()}` }),
     ...(getStore() && { store: getStore() }),
     ...(getCurrency() && { 'Content-Currency': getCurrency() }),
```
[the remark this answers is not in the document]
Comment (sequensucks) [reply packages/api-client/src/api/getHeaders.ts:10]: 2 cases. First, we give users to change all headers. Second, we don't give users to change required headers.

[the remark this answers is not in the document]
Comment (bartoszherba) [reply packages/api-client/src/api/getHeaders.ts:10]: Exactly the opposite.
```
const obj = {a: true, b: true};
const custom = { a: false };
console.log('case1', {...custom, ...obj});
console.log('case2', {...obj, ...custom});

"case1", {
  a: true,
  b: true
}
"case2", {
  a: false,
  b: true
}
```

[the remark this answers is not in the document]
Comment (sequensucks) [reply packages/api-client/src/api/getHeaders.ts:10]: Yes, if ...customHeaders will be first, we don't give users to change required headers. If i change it, users will be able to change all headers.

[the remark this answers is not in the document]
Comment (bartoszherba) [reply packages/api-client/src/api/getHeaders.ts:10]: This is why we introduce custom headers, to allow users to change headers. Do you see any risks? I do not. Each default header can be easily manipulated in another way.

[the remark this answers is not in the document]
Comment (sequensucks) [reply packages/api-client/src/api/getHeaders.ts]: fixed it

File: packages/api-client/src/api/getHeaders.ts [hunk 3]
```diff
@@ -1,9 +1,10 @@
 import { Context } from '../types/context';
 
-export default function getHeaders(context: Context) {
+export default function getHeaders(context: Context, customHeaders: Record<string, string> = {}) {
```
[the remark this answers is not in the document]
Comment (sequensucks) [reply packages/api-client/src/api/getHeaders.ts]: fixed it.

**Review by bartoszherba — APPROVED** [review]

**Pull request merged**